sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
+      }
+
       @Override
       public PCollection<TableRow> expand(PBegin input) {
-        stepUuid = randomUUIDString();
+        final Pipeline p = input.getPipeline();
+        String jobUuid = randomUUIDString();
         BigQueryOptions bqOptions = input.getPipeline().getOptions().as(BigQueryOptions.class);
-        jobUuid = NestedValueProvider.of(
-           StaticValueProvider.of(bqOptions.getJobName()), new CreatePerBeamJobUuid(stepUuid));
-        final ValueProvider<String> jobIdToken = NestedValueProvider.of(
-            jobUuid, new BeamJobUuidToBigQueryJobUuid());
-
-        BoundedSource<TableRow> source;
         final BigQueryServices bqServices = getBigQueryServices();
-
+        final String executingProject = bqOptions.getProject();
         final String extractDestinationDir;
         String tempLocation = bqOptions.getTempLocation();
         try {
           IOChannelFactory factory = IOChannelUtils.getFactory(tempLocation);
-          extractDestinationDir = factory.resolve(tempLocation, stepUuid);
+          extractDestinationDir = factory.resolve(tempLocation, jobUuid);
         } catch (IOException e) {
           throw new RuntimeException(
               String.format("Failed to resolve extract destination directory in %s", tempLocation));
         }
 
-        final String executingProject = bqOptions.getProject();
-        if (query != null && (!query.isAccessible() || !Strings.isNullOrEmpty(query.get()))) {
-          source = BigQueryQuerySource.create(
-              jobIdToken, query, NestedValueProvider.of(
-                jobUuid, new CreateJsonTableRefFromUuid(executingProject)),
-              flattenResults, useLegacySql, extractDestinationDir, bqServices);
+        final PCollectionView<String> jobIdTokenView;
+        PCollection<String> jobIdTokenCollection;
+        if (true) {
+          // Create a singleton job ID token at construction time.
+          jobIdTokenView = p
+              .apply("TriggerIdCreation", Create.of(jobUuid))
+              .apply(View.<String>asSingleton());
         } else {
-          ValueProvider<TableReference> inputTable = getTableWithDefaultProject(bqOptions);
-          source = BigQueryTableSource.create(
-              jobIdToken, inputTable, extractDestinationDir, bqServices,
-              StaticValueProvider.of(executingProject));
+          // Create a singleton job ID token at execution time.
+          jobIdTokenCollection = p
+            .apply("TriggerIdCreation", Create.of("ignored"))
+            .apply("CreateJobId", MapElements.via(
+                new SimpleFunction<String, String>() {
+                  @Override
+                  public String apply(String input) {
+                    return randomUUIDString();
+                  }
+                }));
+          jobIdTokenView = jobIdTokenCollection
+              .apply(View.<String>asSingleton());
         }
+
+        PCollection<TableRow> rows;
+        if (true) {
```

Comment:
What will the condition actually be?.. The non-Source version, without SDF, will not have dynamic rebalancing, so it better be a condition that justifies that :)